framework/include/serializer.hpp

```diff
@@ -0,0 +1,321 @@
+#pragma once
+#include "gvk.hpp"
+/** cereal binary archive */
+#include "cereal/cereal.hpp"
+#include "cereal/archives/binary.hpp"
+
+/** cereal stdlib type support */
+#include "cereal/types/array.hpp"
+#include "cereal/types/atomic.hpp"
+#include "cereal/types/base_class.hpp"
+#include "cereal/types/bitset.hpp"
+#include "cereal/types/chrono.hpp"
+#include "cereal/types/common.hpp"
+#include "cereal/types/complex.hpp"
+#include "cereal/types/deque.hpp"
+#include "cereal/types/forward_list.hpp"
+#include "cereal/types/functional.hpp"
+#include "cereal/types/list.hpp"
+#include "cereal/types/map.hpp"
+#include "cereal/types/memory.hpp"
+#include "cereal/types/optional.hpp"
+#include "cereal/types/polymorphic.hpp"
+#include "cereal/types/queue.hpp"
+#include "cereal/types/set.hpp"
+#include "cereal/types/stack.hpp"
+#include "cereal/types/string.hpp"
+#include "cereal/types/tuple.hpp"
+#include "cereal/types/unordered_map.hpp"
+#include "cereal/types/unordered_set.hpp"
+#include "cereal/types/utility.hpp"
+#include "cereal/types/valarray.hpp"
+#include "cereal/types/variant.hpp"
+#include "cereal/types/vector.hpp"
+
+namespace gvk {
+
+	/** @brief Serializer
+	 *  
+	 *  This type serializes/deserializes objects to/from binary files using the cereal
+	 *  serialization library. The serializer must be initialised either with a
+	 *  serializer::serialize or serializer::deserialize object to handle serialization.
+	 */
+	class serializer
+	{
+	public:
+
+		class serialize {
+			std::ofstream mOfstream;
+			cereal::BinaryOutputArchive mArchive;
+
+		public:
+			serialize(const std::string& aPath) :
+				mOfstream(aPath, std::ios::binary),
+				mArchive(mOfstream)
+			{}
+
+			serialize(serialize&& aOther) noexcept :
+				mOfstream(std::move(aOther.mOfstream)),
+				mArchive(mOfstream)
+			{}
+
+			template<typename Type>
+			void operator()(Type&& aValue)
+			{
+				mArchive(std::forward<Type>(aValue));
+			}
+		};
+
+		class deserialize
+		{
+			std::ifstream mIfstream;
+			cereal::BinaryInputArchive mArchive;
+
+		public:
+			deserialize(const std::string& aPath) :
+				mIfstream(aPath, std::ios::binary),
+				mArchive(mIfstream)
+			{}
+
+			deserialize(deserialize&& aOther) noexcept :
+				mIfstream(std::move(aOther.mIfstream)),
+				mArchive(mIfstream)
+			{}
+
+			template<typename Type>
+			void operator()(Type&& aValue)
+			{
+				mArchive(std::forward<Type>(aValue));
+			}
+		};
+
+		serializer(serialize&& aMode) :
+			mArchive(std::move(aMode)),
+			mMode(mode::serialize)
+		{}
+
+		serializer(deserialize&& aMode) :
+			mArchive(std::move(aMode)),
+			mMode(mode::deserialize)
+		{}
+
+		enum class mode {
+			serialize,
+			deserialize,
+			none
```

Comment:
I think, `none` is not used anywhere? Does it have some use case? If not => remove it.